src/jit/morph.cpp

```diff
@@ -4963,11 +4963,17 @@ GenTreePtr Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPtr f
         objClass           = argObj->gtClass;
         structSize         = info.compCompHnd->getClassSize(objClass);
 
-        // If we have a GT_OBJ of a GT_ADDR then we set argValue to the child node of the GT_ADDR
-        //
-        if (argObj->gtOp1->OperGet() == GT_ADDR)
+        // If we have a GT_OBJ of a GT_ADDR then we set argValue to the child node of the GT_ADDR.
+        GenTree* op1 = argObj->gtOp1;
+        if (op1->OperGet() == GT_ADDR)
         {
-            argValue = argObj->gtOp1->gtOp.gtOp1;
+            GenTree* underlyingTree = op1->gtOp.gtOp1;
+
+            // Only update to the same type.
+            if (underlyingTree->TypeGet() == argValue->TypeGet())
```

Comment:
We have the assert that checks that before `argValue` was struct: 
```
4922 assert(varTypeIsStruct(arg->TypeGet()));
```
and after we still have struct: 
```
5074 // We should still have a TYP_STRUCT
5975  assert(varTypeIsStruct(argValue->TypeGet()));
```

so do not volatile this in the middle, actually it is the only place where we can change  `argValue` value of its type.